Editor: Add search functionality to template swapping modal

yogeshbhutkar opened the pull request:
## What?
Closes #69659

This PR introduces search functionality to the template selection modal, making it easier to find templates efficiently.

## Why?

When dealing with a large number of templates, filtering by title makes selection much more convenient.

## How?

These changes are aligned with the search functionality in the `pattern selection` dialog to maintain consistency.

## Testing Instructions

1. Make sure you have at least 15 templates that can be applied to a page.
2. Create a new page.
3. From the page sidebar, click on `Template`, then `Change template`.
4. Test the search functionality.

## Screencast

https://github.com/user-attachments/assets/3eb9ab71-fa43-412b-8a96-722ba6261fbf

yogeshbhutkar committed:
feat: add search functionality to template swapping modal

yogeshbhutkar committed:
refactor: reuse `searchPatterns` to search `templatesAsPatterns`

yogeshbhutkar commented on packages/block-library/src/index.js:
```diff
@@ -334,3 +334,4 @@ export const __experimentalRegisterExperimentalCoreBlocks =
 		: undefined;
 
 export { privateApis } from './private-apis';
+export { searchPatterns as searchItemsByTitle } from './utils/search-patterns';
```
Since the logic for searching templates closely mirrors that of `searchPatterns`, I’ve opted to reuse the existing code. However, I’m open to suggestions for better approaches or best practices.

Mamaduka replied:
It's probably better to duplicate the code instead of making this a public API.

yogeshbhutkar replied:
Agreed. Thanks 🚀

yogeshbhutkar committed:
feat: add block-library path to editor tsconfig

yogeshbhutkar committed:
feat: add block-library dependency to package-lock.json

Mamaduka commented on packages/editor/package.json:
```diff
@@ -38,6 +38,7 @@
 		"@wordpress/api-fetch": "file:../api-fetch",
 		"@wordpress/blob": "file:../blob",
 		"@wordpress/block-editor": "file:../block-editor",
+		"@wordpress/block-library": "file:../block-library",
```
Suggestion (blocking): Let's not add the whole package as a dependency just for a single util method.

Mamaduka commented on packages/editor/src/components/post-template/swap-template-button.js:
```diff
@@ -4,19 +4,21 @@
 import { useAvailableTemplates, useEditedPostContext } from './hooks';
 
-export default function SwapTemplateButton( { onClick } ) {
+export default function SwapTemplateButton( { onClick, showSearch = true } ) {
```
Why do we need the `showSearch` prop?

yogeshbhutkar replied:
I explored the pattern selection codebase and initially thought `showSearch` might be useful later. But after couple of recent reviews, I learned to focus only on necessary changes. 😅

This was fixed in the latest commit, thanks for the review @Mamaduka.

yogeshbhutkar committed:
revert: remove `block-library` dependency usage

yogeshbhutkar committed:
fix: remove the usage of `showSearch` prop

yogeshbhutkar committed:
fix: correct typo in search template documentation

yogeshbhutkar commented:
![Screenshot 2025-03-25 at 11 04 25 AM](https://github.com/user-attachments/assets/ef795f4c-18a2-4d37-9343-44f9110ad184)

On a separate note, I think it's best to display a simple "No templates/patterns found" message in these cases.

yogeshbhutkar commented:
Hi, @Mamaduka. I believe I've addressed all the feedback. Whenever you have a chance, could you give it another review? Thanks!

Mamaduka commented on packages/editor/src/components/post-template/swap-template-button.js:
```diff
@@ -49,9 +51,18 @@ export default function SwapTemplateButton( { onClick } ) {
 					isFullScreen
 				>
 					<div className="editor-post-template__swap-template-modal-content">
+						<SearchControl
+							__nextHasNoMarginBottom
+							onChange={ setSearchValue }
+							value={ searchValue }
+							label={ __( 'Search' ) }
+							placeholder={ __( 'Search' ) }
+							className="editor-post-template__swap-template-search"
+						/>
```
Currently, the `searchValue` state doesn't reset after you close the modal. I think we need to move the search control inside the + its state inside the `TemplatesList` or create a new component that wraps around both. This way, the local state will reset when the modal component unmounts.

## Screencast 


https://github.com/user-attachments/assets/88b02bdd-6354-41b7-9b7f-fabb6032097c

yogeshbhutkar replied:
Sorry, I didn't notice this in my testing. I’ve factored it in and updated the PR in the latest commit. Thanks for pointing it out!

yogeshbhutkar committed:
fix: move `SearchControl` inside `TemplatesList`

This fixes a bug where the search state was getting preserved even after closing the modal.

Mamaduka reviewed (approved):
Thank you, @yogeshbhutkar!

carolinan commented:
Thank you!
